rocksdb_admin/admin_handler.cpp

```diff
@@ -929,6 +974,177 @@ void AdminHandler::async_tm_addS3SstFilesToDB(
+
+  // Find the watcher for the given path. If none exists, create a new watcher
+  std::shared_ptr<::kafka::KafkaBrokerFileWatcher> kafka_broker_file_watcher;
+
+  const auto map_iter = serverset_path_to_file_watcher_map.find(
+      kafka_broker_serverset_path);
+
+  if (map_iter == serverset_path_to_file_watcher_map.end()) {
+    kafka_broker_file_watcher =
+        std::make_shared<::kafka::KafkaBrokerFileWatcher>(
+            kafka_broker_serverset_path);
+    serverset_path_to_file_watcher_map[kafka_broker_serverset_path] =
+        kafka_broker_file_watcher;
+  } else {
+    kafka_broker_file_watcher = map_iter->second;
+  }
+
+  // Create new synchronization variables for this partition
+  auto sync_variable = std::make_shared<SyncVariables>();
+  sync_var_map[db_name] = sync_variable;
+
+  SCOPE_EXIT { sync_var_map.erase(db_name); };
+
+  const auto kafka_consumer_pool = std::make_shared<::kafka::KafkaConsumerPool>(
+      kKafkaConsumerPoolSize,
+      partition_ids_set,
+      kafka_broker_file_watcher->GetKafkaBrokerList(),
+      std::unordered_set<std::string>({topic_name}),
+      getConsumerGroupId(topic_name, std::to_string(partition_id)),
+      kKafkaConsumerType);
+
+  const auto kafka_watcher = std::make_unique<KafkaWatcher>(
+      kKafkaWatcherName,
+      kafka_consumer_pool,
+      -1); // kafka_init_blocking_consumer_timeout_ms ;
+
+  // With kafka_init_blocking_consumer_timeout_ms set to -1, messages from
+  // replay_timestamp_ms to the current are synchronously consumed. The
+  // calling thread then returns after spawning a new thread to consume
+  // live messages
+  kafka_watcher->StartWith(
+      replay_timestamp_ms,
+      [](std::shared_ptr<const RdKafka::Message> message,
+          const bool is_replay) {
+    if (message == nullptr) {
+      LOG(ERROR) << "Message nullptr";
+      return;
+    }
+    const int64_t msg_timestamp_secs = GetMessageTimestampSecs(*message);
+
+    // Logs for debugging
+    LOG(ERROR) << "Key " << *message->key() << ", "
+    << "value " << static_cast<const char *>(message->payload()) << ", "
+    << "partition: " << message->partition() << ", "
+    << "offset: " << message->offset() << ", "
+    << "payload len: " << message->len() << ", "
+    << "msg_timestamp: " << ToUTC(msg_timestamp_secs) << " or "
+    << std::to_string(msg_timestamp_secs) << " secs";
+
+    // TODO: add the logic to write the key/value to rocksdb here
+
+    // TODO: periodically write the message timestamp to local_meta_db
+  });
+
+  LOG(INFO) << "Now consuming live messages for " << db_name;
+
+  // Live messages continue to consumed in a separate thread, but release
+  // callback so that helix can transition this partition to bootstrap stage.
+  callback.release()->result(StartMessageIngestionResponse());
+
+  // Consume live messages until 'Stop' API is called which will notify the
+  // global condition variable.
+  LOG(INFO) << "Consuming until notified for  " << db_name;
+  std::mutex mutex;
```

Comment:
You would have to put this mutex into SyncVariables also, because both thread need to take this lock